Fallback to OSM if other engines aren't available

seadowg opened the pull request:
This will make development and testing with debug builds (that might not include Mapbox/Google Maps) easier. The app will now fallback to whatever mapping engine is available on start up.

I also moved all Google Maps into a dedicated module (like we already had for OSM and Mapbox), to make it easier for me to keep track of the various components in play.

For @getodk/testers: You should now be able to do test CI builds without Google Maps/Mapbox configured and not have to change to OSM (which should be selected automatically) so that's worth verifying. It's also worth verifying that local builds with Google Maps/Mapbox configured still work as expected. As @grzesiek2010 pointed out, scanning a QR code with Google Maps/Mapbox configured on a build that doesn't have it available will still select that map engine and probably cause problems.

seadowg committed:
Move Google Maps code to new module

seadowg committed:
Only allow Map engines that are available

Fallback to available ones on app start

grzesiek2010 commented on google-maps/src/main/java/org/odk/collect/googlemaps/GoogleMapFragment.java:
```diff
@@ -84,6 +87,28 @@ public class GoogleMapFragment extends SupportMapFragment implements
+                .getDeviceConfigurationInfo().reqGlEsVersion >= 0x20000;
+    }
+
+    private static boolean isGooglePlayServicesAvailable(Context context) {
```
I think this method is redundant and you can call `new PlayServicesChecker().isGooglePlayServicesAvailable(context)` directly above the naming is the same so it's not confusing or something.

seadowg replied:
This was just to to make the line more readable. I could replace it with a local variable in `isAvailable` if you'd prefer that?

grzesiek2010 replied:
Let's leave it. If it's more readable for you that's fine.

grzesiek2010 commented on google-maps/src/main/java/org/odk/collect/googlemaps/GoogleMapFragment.java:
```diff
@@ -84,6 +87,28 @@ public class GoogleMapFragment extends SupportMapFragment implements
+            ApplicationInfo applicationInfo = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
+            String apiKey = applicationInfo.metaData.getString("com.google.android.geo.API_KEY");
+
+            return isGoogleMapsSdkAvailable(context) && isGooglePlayServicesAvailable(context) && !apiKey.equals("");
```
What if the key is incorrect? We can't check it so it would require trying it out and catching exceptions maybe?

seadowg replied:
If the key is incorrect you'd go back to the blank screen case (like we have now if there's no key). That seems fine to me?

grzesiek2010 replied:
Yeah probably... even if it's possible to check if the key is correct by for example trying to display a map and catching exceptions it's not worth it I think.

grzesiek2010 commented on google-maps/src/main/java/org/odk/collect/googlemaps/GoogleMapFragment.java:
```diff
@@ -84,6 +87,28 @@ public class GoogleMapFragment extends SupportMapFragment implements
     // Bundle keys understood by applyConfig().
     static final String KEY_MAP_TYPE = "MAP_TYPE";
 
+    public static boolean isAvailable(Context context) {
```
Don't you think it would be better to keep these methods in a separate class than in the fragment?

seadowg replied:
Ah yeah! This should probably get inlined into `GoogleMapConfigurator`.

grzesiek2010 commented on google-maps/consumer-rules.pro:
If it's empty I would remove it.

seadowg replied:
We've tended to keep `consumer-rules.pro` for all our modules so we can keep the config in `build.gradle` and make it easy to add rules if we need. The file is still being "used" even it's empty.

grzesiek2010 replied:
But why do we need it at all? We have a bunch of such files and all of them are empty.

seadowg replied:
I don't have strong feelings on getting rid of them vs keeping them as it's easy to add one back in if we need it. Do you want to just kill them all as part of this PR?

grzesiek2010 replied:
No it doesn't have to be in this pr but I used to remove them from new modules that I created and since we have more and more of them and they are not used maybe there is no point in adding them and we can remove the existing ones?

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/application/initialization/ApplicationInitializer.kt:
```diff
@@ -97,6 +99,18 @@ class ApplicationInitializer(
     }
 
     private fun initializeMapFrameworks() {
+        // Reset basemap setting if the currently selected is not available
```
If we install the app this code will be executed but what if we scan a QR code with maps that are not available?

seadowg replied:
Ah good point. The QR code map would be set and used (even if it's not available), but it would be then be reset the next time the app is launched (as a fresh process). This could cause a crash I'd imagine if you used a QR code with Mapbox on a build without Mapbox and would give you the blank screen case for Google Maps (if there's no Google Maps key). I'm not sure whether this is worth fixing seeing as we expect all maps to be available in full release builds. What do you think?

grzesiek2010 replied:
Can't we just run that code or a part of it after loading new settings to update maps?

seadowg replied:
Yeah I guess we can pull this code to something shared and call it on settings change as well.

grzesiek2010 replied:
So we can have a separate issue for that improvement? We can ask the QA team to verify this case to make sure it works in the way we assume and then file an issue.

seadowg replied:
Yeah it's actually a little awkward to get that working right now as we'd need to rework things so that `CollectSettingsChangeHandler` is able to actually change settings. I reckon we leave it for now and come back for it it gets annoying.

I'll add a note for QA.

seadowg committed:
Move isAvailable implementation

seadowg committed:
Remove unuse consumer-rules

seadowg committed:
Optimize imports

dbemke commented:
The default map shown is Google (when a map is opened) but in the settings OSM is set. The maps change to OSM only after going to Settings -> Maps.

Steps to reproduce:
1. Install the CircleCI apk and open the demo project.
2. Download forms.
3. Open any map (any geo widget or the form map) and check the source of maps.
4. Go to Settings-> Maps (no need to open the source of maps).
5. Open any map.

seadowg commented:
> The default map shown is Google (when a map is opened) but in the settings OSM is set. The maps change to OSM only after going to Settings -> Maps.

Ah! It looks like it doesn't work for a fresh install when there are no projects as there's none to reset back to OSM. I'll fix that.

seadowg committed:
Reset map base framework when setting/changing project

seadowg committed:
Rename class

This is starting to look and function like FormsDataService, so
it makes sense to make it official.

seadowg committed:
Fix imports

seadowg committed:
Ignore flakey tests

seadowg committed:
Only initalize frameworks once per process

seadowg commented:
@getodk/testers this is good for another look!

srujner commented:
**Tested with Success**

Verified on device with Android 13

Verified cases:
- OSM is selected by default;
- Regression check on maps with OSM;
- Changing Project;
- Changing Map source;
- Offline Layers;
- Scanning a QR code with Google Maps/Mapbox;
- Local builds with Google Maps/Mapbox configured

dbemke commented:
**Tested with Success**

Verified on device with Android 10
